vnext/JSI/Shared/ChakraRuntime.cpp

```diff
@@ -64,132 +67,150 @@ std::vector<JsValueRef> ConstructJsFunctionArguments(
+  // using thread local storage for context data. This is OK for now since we
+  // currently assume that ChakraRuntime will be created and used exclusively in
+  // a single thread, but we need to revisit this if our assumption is wrong.
   VerifyChakraErrorElseThrow(JsSetCurrentContext(m_context));
 
-  startDebuggingIfNeeded();
+  // We evaluate the bootstraping bundle before starting debugging. This is
+  // because ChakraCoreRuntime has a debuggerBreakOnNextLine setting, and we
+  // only want the runtime to break on JS code loaded by the clients.
+  EvaluateJavaScriptSimple(
+      std::make_shared<facebook::jsi::StringBuffer>(g_bootstrapBundleSource), "ChakraRuntime_bootstrap.bundle");
 
-  setupNativePromiseContinuation();
-
-  std::call_once(s_runtimeVersionInitFlag, initRuntimeVersion);
+  if (m_args->enableDebugging) {
+    StartDebugging();
+  }
 
-  static facebook::jsi::StringBuffer bootstrapBundleSourceBuffer{g_bootstrapBundleSource};
-  evaluateJavaScriptSimple(bootstrapBundleSourceBuffer, "ChakraRuntime_bootstrap.bundle");
+  InitializeAdditionalFeatures();
 }
 
 ChakraRuntime::~ChakraRuntime() noexcept {
-  stopDebuggingIfNeeded();
+  // Note: We call VerifyChakraErrorElseCrash instead of
+  // VerifyChakraErrorElseThrow here to prevent any exceptions from leaving this
+  // destrucotr.
+
+  VerifyChakraErrorElseCrash(JsSetCurrentContext(JS_INVALID_REFERENCE));
 
-  VerifyChakraErrorElseThrow(JsSetCurrentContext(JS_INVALID_REFERENCE));
+  // Note: We invalidate m_context explicitly here to avoid calling JsRelease
+  // after the m_runtime has been disposed. Invalidate() can throw, but since we
+  // marked this destructor as noexcept, std::terminate() will be invoked.
   m_context.Invalidate();
 
-  JsSetRuntimeMemoryAllocationCallback(m_runtime, nullptr, nullptr);
+  VerifyChakraErrorElseCrash(JsSetRuntimeMemoryAllocationCallback(m_runtime, nullptr, nullptr));
 
-  VerifyChakraErrorElseThrow(JsDisposeRuntime(m_runtime));
+  VerifyChakraErrorElseCrash(JsDisposeRuntime(m_runtime));
 }
 
 #pragma region Functions_inherited_from_Runtime
 
 facebook::jsi::Value ChakraRuntime::evaluateJavaScript(
     const std::shared_ptr<const facebook::jsi::Buffer> &buffer,
-    const std::string &sourceURL) {
-  // Simple evaluate if scriptStore not available as it's risky to utilize the
-  // byte codes without checking the script version.
-  if (!runtimeArgs().scriptStore) {
-    if (!buffer)
-      throw facebook::jsi::JSINativeException("Script buffer is empty!");
-    return evaluateJavaScriptSimple(*buffer, sourceURL);
+    const std::string &sourceUrl) {
+  if (!buffer) {
+    throw facebook::jsi::JSINativeException("Error in ChakraRuntime::evaluateJavaScript() - buffer cannot be null.");
   }
 
-  uint64_t scriptVersion = 0;
-  std::shared_ptr<const facebook::jsi::Buffer> scriptBuffer;
+  // PreparedJavaScript optimization is disabled when debugging.
+  if (m_args->scriptStore && m_args->preparedScriptStore && !m_args->enableDebugging) {
```

Comment:
>scriptStore [](start = 14, length = 11)

Should we be doing anything with the scriptstore here? We have a separate evaluatePreparedJavaScript method on the Runtime. Should a higher level of abstraction be the one choosing the right thing to do?

Reply:
In general though, I think we could do a better job of separating changes still. It's a bit hard in this CR to tell what is a mechanical move compared to new logic which seems independent of the class hierarchy changes and should get some more scrutiny.

Smaller PRs are going to be easier to review and catch mistakes in, but it's also useful from the perspective of being able to revert individual changes instead of a whole thing.

Would you be willing to spend some time to separate mechanical + non-mechanical, and consider having a PR be a couple classes at a time? It feels like we can take smaller chunks of this CR and still build and run correctly.

---
In reply to: [349395567](https://github.com/microsoft/react-native-windows/pull/3677#discussion_r349395567) [](ancestors = 349395567)

Reply:
I won't be able to work on this tomorrow, but I'd like to respond to the general comment about breaking this change down further. I completely understand the appeal of smaller PRs, but I am not sure how much more I can do here. I believe that iterations 3 through 6 have to go together, as they consist of the actual class hierarchy implementation and the required downstream changes. I can make Iterations 1 & 2 separate PRs if that helps, though I would point out that checking in iteration 1 as a separate commit will leave us in a state where, while there are files named CharkraCoreRuntime.cpp and ChakraRtRuntime.cpp, CharkraCoreRuntime and ChakraRtRuntime do not exist in code. Please let me know what you think.

---
In reply to: [349396678](https://github.com/microsoft/react-native-windows/pull/3677#discussion_r349396678) [](ancestors = 349396678,349395567)

Reply:
From a high level view, it feels like I can divide some of the changes in this CR into
1. Changes to naming/capitalization/namespaces (which add scattered diffs)
2. Non-trivial changes to logic, like the above
3. Tweaks or new documentation for existing APIs
4. More constrainable changes, like moves or fixes needed to abstract away functionality to subclasses

From what I can tell, these categories of changes are mostly independent/atomic, or at least could mostly be done without the others. Even for things like pulling a class into subclasses, we could divide changes into methods at a time if we wanted to.

I'm not sure what the best granularity would be here, but I when I see CRs get this large, things usually get missed. I'm personally struggling a bit to contextualize some of these changes against each other. It might be some work to divide this, but I think it could be really useful in order to generate some more productive feedback on the changes.

---
In reply to: [349404439](https://github.com/microsoft/react-native-windows/pull/3677#discussion_r349404439) [](ancestors = 349404439,349396678,349395567)